src/frontend/data/stores/useChatItemsStore/index.ts

```diff
@@ -0,0 +1,148 @@
+}
+
+export interface ChatPresenceItemWrapper {
+  type: chatItemType.PRESENCE;
+  presenceData: ChatPresenceType;
+}
+
+export type ChatItem = ChatMessageGroupWrapper | ChatPresenceItemWrapper;
+
+type State = {
+  hasReceivedMessageHistory: boolean;
+  setHasReceivedMessageHistory: (received: boolean) => void;
+  chatItems: ChatItem[];
+  addMessage: (newMessage: ReceivedMessageType) => void;
+  addPresence: (newPresence: ChatPresenceType) => void;
+};
+
+export const useChatItemState = create<State>((set) => ({
+  hasReceivedMessageHistory: false,
+  setHasReceivedMessageHistory: (received) =>
+    set({
+      hasReceivedMessageHistory: received,
+    }),
+  chatItems: [],
+  addMessage: (newMessage: ReceivedMessageType) =>
+    set((state) => {
+      // Get the chatMessageGroup where new message will be added (if group exists)
+      const correspondingChatMessageGroup =
+        getChatMessageGroupForNewReceivedMessage(state.chatItems, newMessage);
+      if (correspondingChatMessageGroup) {
+        {
+          return {
+            chatItems: state.chatItems.map((chatItem) => {
+              if (
+                chatItem.type === chatItemType.GROUP_MESSAGE &&
+                chatItem.messageGroupData === correspondingChatMessageGroup
+              ) {
+                return {
+                  type: chatItemType.GROUP_MESSAGE,
+                  messageGroupData: {
+                    sender: correspondingChatMessageGroup.sender,
+                    messages: [
+                      ...chatItem.messageGroupData.messages,
+                      {
+                        content: newMessage.content,
+                        sentAt: newMessage.sentAt,
+                      },
+                    ],
+                  },
+                };
+              }
+              return chatItem;
+            }),
+          };
+        }
+      }
+
+      return {
+        chatItems: state.chatItems.concat({
+          type: chatItemType.GROUP_MESSAGE,
+          messageGroupData: {
+            messages: [
+              {
+                content: newMessage.content,
+                sentAt: newMessage.sentAt,
+              },
+            ],
+            sender: newMessage.sender,
+          },
+        }),
+      };
+    }),
+  addPresence: (presence) =>
+    set((state) => ({
+      chatItems: state.chatItems.concat({
+        type: chatItemType.PRESENCE,
+        presenceData: presence,
+      }),
+    })),
+}));
+
+const getChatMessageGroupForNewReceivedMessage = (
+  chatItems: ChatItem[],
+  newReceivedMessage: ReceivedMessageType,
+): ChatMessageGroupType | null => {
+  // Retrieve only chatMessagesGroup
+  const groupMessages = chatItems.map((chatItem) => {
+    if (chatItem.type === chatItemType.GROUP_MESSAGE) {
+      return chatItem;
+    }
+  });
+  if (groupMessages && groupMessages.length) {
+    const lastGroupMessage =
+      groupMessages[groupMessages.length - 1]?.messageGroupData;
```

Comment:
I suggested you to test as an invariant that a group message list can not be empty then you will be able to remove this optional (and eventually add a comment here for the reason why you can do it)